In [49]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import test
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 9: Rope Bridge ---</h2><p>This rope bridge creaks as you walk along it. You aren't sure how old it is, or whether it can even support your weight.</p>
<p>It seems to support the Elves just fine, though. The bridge spans a gorge which was carved out by the massive river far below you.</p>
<p>You step carefully; as you do, the ropes stretch and twist. You decide to distract yourself by modeling rope physics; maybe you can even figure out where <em>not</em> to step.</p>
<p>Consider a rope with a knot at each end; these knots mark the <em>head</em> and the <em>tail</em> of the rope. If the head moves far enough away from the tail, the tail is pulled toward the head.</p>
<p>Due to nebulous reasoning involving <a href="https://en.wikipedia.org/wiki/Planck_units#Planck_length" target="_blank">Planck lengths</a>, you should be able to model the positions of the knots on a two-dimensional grid. Then, by following a hypothetical <em>series of motions</em> (your puzzle input) for the head, you can determine how the tail will move.</p>
<p><span title="I'm an engineer, not a physicist!">Due to the aforementioned Planck lengths</span>, the rope must be quite short; in fact, the head (<code>H</code>) and tail (<code>T</code>) must <em>always be touching</em> (diagonally adjacent and even overlapping both count as touching):</p>
<pre><code>....
.TH.
....

....
.H..
..T.
....

...
.H. (H covers T)
...
</code></pre>

<p>If the head is ever two steps directly up, down, left, or right from the tail, the tail must also move one step in that direction so it remains close enough:</p>
<pre><code>.....    .....    .....
.TH.. -&gt; .T.H. -&gt; ..TH.
.....    .....    .....

... ... ...
.T. .T. ...
.H. -&gt; ... -&gt; .T.
... .H. .H.
... ... ...
</code></pre>

<p>Otherwise, if the head and tail aren't touching and aren't in the same row or column, the tail always moves one step diagonally to keep up:</p>
<pre><code>.....    .....    .....
.....    ..H..    ..H..
..H.. -&gt; ..... -&gt; ..T..
.T...    .T...    .....
.....    .....    .....

..... ..... .....
..... ..... .....
..H.. -&gt; ...H. -&gt; ..TH.
.T... .T... .....
..... ..... .....
</code></pre>

<p>You just need to work out where the tail goes as the head follows a series of motions. Assume the head and the tail both start at the same position, overlapping.</p>
<p>For example:</p>
<pre><code>R 4
U 4
L 3
D 1
R 4
D 1
L 5
R 2
</code></pre>
<p>This series of motions moves the head <em>right</em> four steps, then <em>up</em> four steps, then <em>left</em> three steps, then <em>down</em> one step, and so on. After each step, you'll need to update the position of the tail if the step means the head is no longer adjacent to the tail. Visually, these motions occur as follows (<code>s</code> marks the starting position as a reference point):</p>
<pre><code>== Initial State ==

......
......
......
......
H..... (H covers T, s)

== R 4 ==

......
......
......
......
TH.... (T covers s)

......
......
......
......
sTH...

......
......
......
......
s.TH..

......
......
......
......
s..TH.

== U 4 ==

......
......
......
....H.
s..T..

......
......
....H.
....T.
s.....

......
....H.
....T.
......
s.....

....H.
....T.
......
......
s.....

== L 3 ==

...H..
....T.
......
......
s.....

..HT..
......
......
......
s.....

.HT...
......
......
......
s.....

== D 1 ==

..T...
.H....
......
......
s.....

== R 4 ==

..T...
..H...
......
......
s.....

..T...
...H..
......
......
s.....

......
...TH.
......
......
s.....

......
....TH
......
......
s.....

== D 1 ==

......
....T.
.....H
......
s.....

== L 5 ==

......
....T.
....H.
......
s.....

......
....T.
...H..
......
s.....

......
......
..HT..
......
s.....

......
......
.HT...
......
s.....

......
......
HT....
......
s.....

== R 2 ==

......
......
.H.... (H covers T)
......
s.....

......
......
.TH...
......
s.....
</code></pre>

<p>After simulating the rope, you can count up all of the positions the <em>tail visited at least once</em>. In this diagram, <code>s</code> again marks the starting position (which the tail also visited) and <code>#</code> marks other positions the tail visited:</p>
<pre><code>..##..
...##.
.####.
....#.
s###..
</code></pre>
<p>So, there are <code><em>13</em></code> positions the tail visited at least once.</p>
<p>Simulate your complete hypothetical series of motions. <em>How many positions does the tail of the rope visit at least once?</em></p>
</article>


In [50]:
class RopeBrigde:
    def __init__(self, number_of_tails: int = 1) -> None:
        self.s = [0, 0]
        self.H = [0, 0]
        self.T = [[0, 0] for _ in range(number_of_tails)]

    def moves(self, s: str, prt=False) -> int:
        moves = [l.strip().split() for l in s.strip().splitlines()]
        tail_has_been = set()

        if prt:
            print("== Initial State ==\n")
            print(self)
            print()

        for direction, distance in moves:

            if prt:
                print(f"== {direction} {distance} ==\n")

            for _ in range(int(distance)):
                self.update_head(direction)
                self.update_tail()

                tail_has_been.add(tuple(self.T[-1]))

                if prt:
                    print(self)
                    print()

        if prt:
            print("Tail positions")
            print(self.tail_position_on_grid_str(tail_has_been))

        return len(tail_has_been)

    def tail_position_on_grid_str(self, tail_has_been) -> str:
        max_rows = max(
            0, max(t[0] for t in self.T), self.H[0], max(r for r, _ in tail_has_been)
        )
        max_cols = max(
            0, max(t[1] for t in self.T), self.H[1], max(c for _, c in tail_has_been)
        )

        min_rows = min(
            -5, min(t[0] for t in self.T), self.H[0], min(r for r, _ in tail_has_been)
        )
        min_cols = min(
            -6, min(t[1] for t in self.T), self.H[1], min(c for _, c in tail_has_been)
        )

        scope_cols = max_cols - min_cols
        scope_rows = max_rows - min_rows

        grid = [["."] * scope_cols for _ in range(scope_rows)]

        for pos in tail_has_been:
            grid[pos[0] - min_rows - 1][pos[1] - min_rows - 1] = "#"

        grid[self.s[0] - min_rows - 1][self.s[1] - min_rows - 1] = "s"
        s = "\n".join("".join(l) for l in grid)
        return s

    def update_tail(self) -> None:
        prev = self.H
        for t in self.T:
            if t[0] == prev[0] and abs(t[1] - prev[1]) > 1:
                t[1] += (prev[1] - t[1]) // abs(prev[1] - t[1])
            elif abs(t[0] - prev[0]) > 1 and t[1] == prev[1]:
                t[0] += (prev[0] - t[0]) // abs(prev[0] - t[0])
            elif abs(t[0] - prev[0]) > 1 or abs(t[1] - prev[1]) > 1:
                t[0] += (prev[0] - t[0]) // abs(prev[0] - t[0])
                t[1] += (prev[1] - t[1]) // abs(prev[1] - t[1])

            prev = t

    def update_head(self, direction: str) -> None:
        if direction == "R":
            self.H[1] += 1
        elif direction == "L":
            self.H[1] -= 1
        if direction == "D":
            self.H[0] += 1
        if direction == "U":
            self.H[0] -= 1

    def __str__(self) -> str:
        max_rows = max(0, max(t[0] for t in self.T), self.H[0])
        max_cols = max(6, max(t[1] for t in self.T), self.H[1])

        min_rows = min(-5, max(t[0] for t in self.T), self.H[0])
        min_cols = min(0, max(t[1] for t in self.T), self.H[1])

        scope_cols = max_cols - min_cols
        scope_rows = max_rows - min_rows

        grid = [["."] * scope_cols for _ in range(scope_rows)]
        grid[self.s[0] - min_rows - 1][self.s[1] + min_cols] = "s"

        for i, t in reversed(list(enumerate(self.T))):
            grid[t[0] - min_rows - 1][t[1] + min_cols] = str(i + 1)

        grid[self.H[0] - min_rows - 1][self.H[1] + min_cols] = "H"
        s = "\n".join("".join(l) for l in grid)

        if self.s == self.H:
            s += f"  (H covers {','.join(str(i+1) for i,t in enumerate(self.T) if t == self.H)}, s)"
        if any(self.H == t for t in self.T):
            s += f"  (H covers {','.join(str(i+1) for i,t in enumerate(self.T) if t == self.H)})"
        if any(self.s == t for t in self.T):
            first, *rest = [i for i, t in enumerate(self.T) if t == self.s]
            s += f"  ({first+1} covers {', '.join(str(i+1) for i in rest)} s)"

        return s


tests = [
    {
        "name": "Example",
        "s": """
            R 4
            U 4
            L 3
            D 1
            R 4
            D 1
            L 5
            R 2
        """,
        "expected": 13,
    },
]


@test(tests=tests)
def test_part_I(s: str) -> int:
    return RopeBrigde().moves(s)


Test Example passed, for test_part_I.
Success


In [51]:
with open("../input/day9.txt") as f:
    puzzle = f.read()

print(f"Part 1: {RopeBrigde().moves(puzzle)}")

Part 1: 6563


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>6563</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>


<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>A rope snaps! Suddenly, the river is getting a lot closer than you remember. The bridge is still there, but some of the ropes that broke are now whipping toward you as you fall through the air!</p>
<p>The ropes are moving too quickly to grab; you only have a few seconds to choose how to arch your body to avoid being hit. Fortunately, your simulation can be extended to support longer ropes.</p>
<p>Rather than two knots, you now must simulate a rope consisting of <em>ten</em> knots. One knot is still the head of the rope and moves according to the series of motions. Each knot further down the rope follows the knot in front of it using the same rules as before.</p>
<p>Using the same series of motions as the above example, but with the knots marked <code>H</code>, <code>1</code>, <code>2</code>, ..., <code>9</code>, the motions now occur as follows:</p>
<pre><code>== Initial State ==

......
......
......
......
H..... (H covers 1, 2, 3, 4, 5, 6, 7, 8, 9, s)

== R 4 ==

......
......
......
......
1H.... (1 covers 2, 3, 4, 5, 6, 7, 8, 9, s)

......
......
......
......
21H... (2 covers 3, 4, 5, 6, 7, 8, 9, s)

......
......
......
......
321H.. (3 covers 4, 5, 6, 7, 8, 9, s)

......
......
......
......
4321H. (4 covers 5, 6, 7, 8, 9, s)

== U 4 ==

......
......
......
....H.
4321.. (4 covers 5, 6, 7, 8, 9, s)

......
......
....H.
.4321.
5..... (5 covers 6, 7, 8, 9, s)

......
....H.
....1.
.432..
5..... (5 covers 6, 7, 8, 9, s)

....H.
....1.
..432.
.5....
6..... (6 covers 7, 8, 9, s)

== L 3 ==

...H..
....1.
..432.
.5....
6..... (6 covers 7, 8, 9, s)

..H1..
...2..
..43..
.5....
6..... (6 covers 7, 8, 9, s)

.H1...
...2..
..43..
.5....
6..... (6 covers 7, 8, 9, s)

== D 1 ==

..1...
.H.2..
..43..
.5....
6..... (6 covers 7, 8, 9, s)

== R 4 ==

..1...
..H2..
..43..
.5....
6..... (6 covers 7, 8, 9, s)

..1...
...H.. (H covers 2)
..43..
.5....
6..... (6 covers 7, 8, 9, s)

......
...1H. (1 covers 2)
..43..
.5....
6..... (6 covers 7, 8, 9, s)

......
...21H
..43..
.5....
6..... (6 covers 7, 8, 9, s)

== D 1 ==

......
...21.
..43.H
.5....
6..... (6 covers 7, 8, 9, s)

== L 5 ==

......
...21.
..43H.
.5....
6..... (6 covers 7, 8, 9, s)

......
...21.
..4H.. (H covers 3)
.5....
6..... (6 covers 7, 8, 9, s)

......
...2..
..H1.. (H covers 4; 1 covers 3)
.5....
6..... (6 covers 7, 8, 9, s)

......
...2..
.H13.. (1 covers 4)
.5....
6..... (6 covers 7, 8, 9, s)

......
......
H123.. (2 covers 4)
.5....
6..... (6 covers 7, 8, 9, s)

== R 2 ==

......
......
.H23.. (H covers 1; 2 covers 4)
.5....
6..... (6 covers 7, 8, 9, s)

......
......
.1H3.. (H covers 2, 4)
.5....
6..... (6 covers 7, 8, 9, s)
</code></pre>

<p>Now, you need to keep track of the positions the new tail, <code>9</code>, visits. In this example, the tail never moves, and so it only visits <code><em>1</em></code> position. However, <em>be careful</em>: more types of motion are possible than before, so you might want to visually compare your simulated rope to the one above.</p>
<p>Here's a larger example:</p>
<pre><code>R 5
U 8
L 8
D 3
R 17
D 10
L 25
U 20
</code></pre>
<p>These motions occur as follows (individual steps are not shown):</p>
<pre><code>== Initial State ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
...........H.............. (H covers 1, 2, 3, 4, 5, 6, 7, 8, 9, s)
..........................
..........................
..........................
..........................
..........................

== R 5 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
...........54321H......... (5 covers 6, 7, 8, 9, s)
..........................
..........................
..........................
..........................
..........................

== U 8 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
................H.........
................1.........
................2.........
................3.........
...............54.........
..............6...........
.............7............
............8.............
...........9.............. (9 covers s)
..........................
..........................
..........................
..........................
..........................

== L 8 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
........H1234.............
............5.............
............6.............
............7.............
............8.............
............9.............
..........................
..........................
...........s..............
..........................
..........................
..........................
..........................
..........................

== D 3 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
.........2345.............
........1...6.............
........H...7.............
............8.............
............9.............
..........................
..........................
...........s..............
..........................
..........................
..........................
..........................
..........................

== R 17 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
................987654321H
..........................
..........................
..........................
..........................
...........s..............
..........................
..........................
..........................
..........................
..........................

== D 10 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
...........s.........98765
.........................4
.........................3
.........................2
.........................1
.........................H

== L 25 ==

..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
...........s..............
..........................
..........................
..........................
..........................
H123456789................

== U 20 ==

H.........................
1.........................
2.........................
3.........................
4.........................
5.........................
6.........................
7.........................
8.........................
9.........................
..........................
..........................
..........................
..........................
..........................
...........s..............
..........................
..........................
..........................
..........................
..........................

</code></pre>

<p>Now, the tail (<code>9</code>) visits <code><em>36</em></code> positions (including <code>s</code>) at least once:</p>
<pre><code>..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
..........................
#.........................
#.............###.........
#............#...#........
.#..........#.....#.......
..#..........#.....#......
...#........#.......#.....
....#......s.........#....
.....#..............#.....
......#............#......
.......#..........#.......
........#........#........
.........########.........
</code></pre>
<p>Simulate your complete series of motions on a larger rope with ten knots. <em>How many positions does the tail of the rope visit at least once?</em></p>
</article>


In [52]:
tests = [
    {
        "name": "Example",
        "s": """
            R 4
            U 4
            L 3
            D 1
            R 4
            D 1
            L 5
            R 2
        """,
        "expected": 1,
    },
    {
        "name": "Example",
        "s": """
            R 5
            U 8
            L 8
            D 3
            R 17
            D 10
            L 25
            U 20
        """,
        "expected": 36,
    },
]


@test(tests=tests)
def test_part_I(s: str) -> int:
    return RopeBrigde(9).moves(s)


Test Example passed, for test_part_I.
Test Example passed, for test_part_I.
Success


<link href="style.css" rel="stylesheet"></link>


In [53]:
print(f"Part 2: {RopeBrigde(9).moves(puzzle)}")

Part 2: 2653


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>2653</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
